DTLS fixes with WANT_WRITE simulations

julek-wolfssl opened the pull request:
- WANT_WRITE could be returned in unexpected places. This patch takes care of that.
- Change state after SendBuffered only if in a sending state to begin with.
- Adapt client and server to simulate WANT_WRITE with DTLS

Reproducible with
```
./udp_proxy -p 12345 -s 127.0.0.1:11111 -S client -x 0:5 # adjust so that the client's finished msg is dropped
./examples/server/server -u -N
./examples/client/client -u -p 12345 -N -6
```

julek-wolfssl committed:
DTLS fixes with WANT_WRITE simulations

- WANT_WRITE could be returned in unexpected places. This patch takes care of that.
- Change state after SendBuffered only if in a sending state to begin with.
- Adapt client and server to simulate WANT_WRITE with DTLS

julek-wolfssl commented on src/tls13.c:
```diff
@@ -8637,9 +8637,18 @@ int wolfSSL_connect_TLSv13(WOLFSSL* ssl)
+                    ssl->options.connectState == FIRST_REPLY_THIRD ||
+                    ssl->options.connectState == FIRST_REPLY_FOURTH) {
+                    ssl->options.connectState++;
+                    WOLFSSL_MSG("connect state: "
```
If we could redesign the implementation I suppose it would be better to set flags to do that but unfortunately that would be a big rewrite. This is just the minimal way to fix the issue.

SparkiDev commented on src/ssl.c:
```diff
@@ -11831,9 +11831,17 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
-                    ssl->options.connectState++;
-                    WOLFSSL_MSG("connect state: "
-                                "Advanced from last buffered fragment send");
+                    if (ssl->options.connectState == CONNECT_BEGIN ||
```
Can we use a range here?

SparkiDev replied:
OK

SparkiDev commented on wolfssl/test.h:
```diff
@@ -1843,10 +1843,8 @@ static WC_INLINE void tcp_connect(SOCKET_T* sockfd, const char* ip, word16 port,
-        if (connect(*sockfd, (const struct sockaddr*)&addr, sizeof(addr)) != 0)
-            err_sys_with_errno("tcp connect failed");
-    }
+    if (connect(*sockfd, (const struct sockaddr*)&addr, sizeof(addr)) != 0)
```
Are you sure we can do this?
We call tcp_connect even when doing DTLS.

julek-wolfssl committed:
Fix resumption failure and use range in connect state logic

ejohnstown reviewed (changes requested):
% ./configure --enable-dtls && make
    % ./examples/server/server -u -v 3 -l DHE-RSA-CHACHA20-POLY1305 -2
    % ./examples/client/client -u -v 3 -l DHE-RSA-CHACHA20-POLY1305 -2

Client fails with a -308, and the unit.test also fails. (Repro on macOS on M1 and x64. No repro on Linux.)

Update 1

`sendto()` is setting errno to 56, `EISCONN`. When I comment out the connect() used on the datagram socket, it just works. This isn't even making it to sending the Client Hello on the Mac. The man page for sendto on the Mac says that sendto can be used on connected sockets.

julek-wolfssl committed:
Reset ret in client and server after wolfSSL_dtls_got_timeout()

- Do UDP connect only with simulateWantWrite to accommodate macOS that doesn't like sendto being called on connected UDP sockets
- Call wolfSSL_dtls_get_current_timeout only on a DTLS connection

julek-wolfssl commented:
@ejohnstown Please run your tests again. The last commit should fix it so that we don't "connect" a UDP socket when we are  going to use `sendto`.
